web-api/src/mind.rs

```diff
@@ -41,6 +45,39 @@ struct Article {
     url: String,
 }
 
+#[derive(Debug, Deserialize)]
+struct ArticleProvider {
+    articles: HashMap<String, Article>,
+}
+
+impl ArticleProvider {
+    fn new(path: &str) -> Result<Self, anyhow::Error> {
+        let mut articles = HashMap::new();
+        for article in read::<Article>(path)? {
+            let article = article?;
+            articles.insert(article.id.clone(), article);
+        }
+        Ok(Self { articles })
+    }
+
+    fn sample(&self, n: usize) -> Vec<Article> {
+        let mut rng = thread_rng();
+        self.articles
+            .values()
+            .cloned()
+            .collect_vec()
+            .iter()
+            .choose_multiple(&mut rng, n)
+            .into_iter()
+            .cloned()
+            .collect()
+    }
```

Comment:
Looks like it could be optimized but didn't find valuable solutions for this, maybe @janpetschexain

Reply:
do we actually need to return owned articles (i guess not, even if we need to own some of the articles later on we could still clone them when needed)? if not we could simplify and save quite a lot by
```rust
    fn sample(&self, n: usize) -> Vec<&Article> {
        self.articles.values().choose_multiple(&mut thread_rng(), n)
    }
```

if we have to return owned articles, then we need to clone the chosen articles, but we can save the first cloning & collecting (this works because `values()` is already an iterator where we can directly apply the iterator extension `choose_multiple()` onto) by
```rust
    fn sample(&self, n: usize) -> Vec<Article> {
        self.articles
            .values()
            .choose_multiple(&mut thread_rng(), n)
            .into_iter()
            .cloned()
            .collect()
    }
```

in any case the handle to the rng could be directly supplied to the iterator instead of first binding it to a variable.